Specify when clients can serve block and sidecars in byRoot RPC methods (#3551)
ethereum/consensus-specs: specs/deneb/p2p-interface.md

No more than `MAX_REQUEST_BLOCKS_DENEB` may be requested at a time.

*[Modified in Deneb:EIP4844]*

> Clients SHOULD include a block in the response as soon as it passes the gossip validation rules.

Reviewer: ```suggestion
Clients SHOULD include a block in the response as soon as it passes the gossip validation rules.
Clients SHOULD NOT respond with blocks that fail the beacon chain state transition.
```

> Clients SHOULD consider including a block in the response as soon as it it passes the gossip validation rules.

Reviewer 2: ```suggestion
Clients SHOULD consider including a block in the response as soon as it passes the gossip validation rules.
```

Clients MUST respond with at least one sidecar, if they have it.
Clients MAY limit the number of blocks and sidecars in the response.

Clients SHOULD include a sidecar in the response as soon as it passes the gossip validation rules.
Clients SHOULD NOT respond with sidecars related to blocks that fail gossip validation rules.

> Clients SHOULD NOT include sidecars related to blocks that fail `state_transition(block)`.

Reviewer 2: ```suggestion
```

> Clients SHOULD include a sidecar in the response as soon as it passes the gossip validation rules.
> Clients SHOULD NOT respond with sidecars related to blocks that fail gossip validation rules.
> Clients SHOULD NOT include sidecars related to blocks that fail `state_transition(block)`.

Reviewer: ```suggestion
Clients SHOULD include a sidecar in the response as soon as it passes the gossip validation rules.
Clients SHOULD NOT respond with sidecars related to blocks that fail gossip validation rules.
Clients SHOULD NOT respond with sidecars related to blocks that fail the beacon chain state transition
```

Clients MUST respond with at least one sidecar, if they have it.
Clients MAY limit the number of blocks and sidecars in the response.

Clients SHOULD include a sidecar in the response as soon as it passes the gossip validation rules.

> Clients MUST NOT respond with sidecars that fails gossip validation rules.

Author: To me this explicitly states that client must not try to serve this before completing the validation, and if validation fails it must not be considered when serving the RPC call.
So there are no assumptions on clients internal RPC and gossip processing.
I agree that probably for most of current clients (teku for sure) this is true by default.

Is there somewhere a general statement that says "messages must pass gossip validation rules before being considered by any mean by the client"?

> Clients MUST NOT respond with sidecars that fails gossip validation rules.

Reviewer: There is not other than the fact that gossipsub generally is gating messages from the application layer on these conditions. Not opposed to a general note if it's warranted, but I don't think we need to call out for each specific instance

> Clients MUST NOT respond with sidecars that fails gossip validation rules.

Author: Fair, let's remove it from here.

> Clients MUST NOT respond with sidecars related to blocks that fail gossip validation rules.

Reviewer 2: ```suggestion
Clients SHOULD NOT respond with sidecars related to blocks that fail gossip validation rules.
```
I believe we can’t use MUST NOT because a block isn’t always around, with MUST NOT a client won’t be able to respond with a sidecar until it pulls a block and runs gossip validation checks on it

Author: ah, I see

> Clients MAY continue serving blocks\sidecars after failed `fork_choice.on_block` if and only if the failure reason is due to missing blobs. (i.e. some sidecars are missing but the available ones have been fully verified against the fully validated block).

Reviewer 2: I think this statement is worth to be duplicated in the `BeaconBlocksByRoot` section
